backbone.js

```diff
@@ -1134,7 +1134,15 @@
       }
       options = options ? _.clone(options) : {};
       options.collection = this;
-      var model = new this.model(attrs, options);
+
+      var model;
+      if (this.model.prototype) {
+        model = new this.model(attrs, options);
+      } else {
+        // ES class methods didn't have prototype
+        model = this.model(attrs, options);
+      }
```

Comment:
This change doesn't make any difference. Try this:

```js
function returnPlainObject() {
    return {};
}

var x = new returnPlainObject();

console.log(x, x instanceof returnPlainObject);
```

Reply:
I'm not quite follow you. Did example from attached issue works?

Reply:
Regarding #4233, I realize I might have jumped to conclusions too early, so I have reopened it. My apologies on that one.

Regarding this review comment, I just meant that your change will not make a difference. If `this.model` is a function that returns an instance of a `Backbone.Model` subclass, then `this.model(...)` does exactly the same thing as `new this.model(...)`. Therefore your change will also not fix #4233. You would need to bring a solid test to convince me otherwise. :-)

Reply:
Ah, I understand now. The "class methods are nonconstructible" problem. Man, ES classes are such a pain.

I would prefer that you add a standalone `construct` function with the same semantics as `new` (but avoiding the keyword), so we can write `var model = construct(this.model, attrs, options)` over here. That function goes like this (Underscore and ES3 syntax for backwards compatibility):

```js
var construct = _.restArguments(function(ctor, args) {
    // _.create works even if the argument is undefined
    var instance = _.create(ctor.prototype);
    var result = ctor.apply(instance, args);
    return _.isObject(result) ? result : instance;
};
```

However, this is not a hard condition for acceptance. The only hard condition is that you add a test, so we can detect it if this breaks again.

Reply:
I'm not sure that i can add test here, as long it involves es6 syntax, even with babel es class methods will likely would be transpiled to `function`

Reply:
You can write a standalone function and then manually set its `prototype` to `undefined`. That should give the same problem with the old code. Probably good to first check that your test fails with the old logic, though.

Reply:
I learned from a quick test that this approach probably won't work:

```js
function test() {}
test.prototype = undefined;
new test; // no error, returns empty object
```

I'm still looking whether I can hack this in with ES3 syntax. Failing that, we can have a test with ES6 syntax, as long as we put it in a separate module (so we can skip it in pre-ES6 environments).

Reply:
Here is a way to create a non-constructor within an ES3 module if the engine supports it, and skip the test otherwise:

```js
// ES6 notation as a string
var esMethod = '{\n' +
'    model(attrs, options) {\n' +
'        return new TestModel(attrs, options);\n' +
'    }\n' +
'}';

(globalThis || self || window).TestModel = Backbone.Model.extend(...);

try {
    var nonConstructible = Function('return ' + esMethod + '.model;')();
    var TestCollection = Backbone.Collection.extend({
        model: nonConstructible,
        modelId: function(attrs) {...}
    });
    // remainder of test here
} catch (error) {
    // syntax not supported, skip test
}
```

Reply:
But this test will always fail, because of lack of support?

Reply:
I see syntax error in Phantom, i'm not sure that my code even invoked

Reply:
The trick is in the `Function(...)()` and the `try`/`catch`. Do you see how that works?

I just realized that you may need to put *only* the `Function(...)()` in the `try` block and return early from the `catch`. Otherwise you will not be able to distinguish the syntax being unsupported from the `TypeError` due to `collection.model` not being constructible.

Reply:
What should test do if there a syntaxerror?

Reply:
It will go into the `catch` block.

Reply:
> I see syntax error in Phantom, i'm not sure that my code even invoked

Oops, missed this comment. Could you show the exact code you're feeding to PhantomJS?

Reply:
Check out my latest commit with test